Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/function/FunctionModule.java

```diff
@@ -0,0 +1,108 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or substantial portions of
+ * the Software.
+ *
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+ * THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.microsoft.tooling.msservices.serviceexplorer.azure.function;
+
+import com.microsoft.azure.CloudException;
+import com.microsoft.azure.management.appservice.FunctionApp;
+import com.microsoft.azuretools.azurecommons.helpers.AzureCmdException;
+import com.microsoft.azuretools.azurecommons.helpers.NotNull;
+import com.microsoft.azuretools.core.mvp.model.ResourceEx;
+import com.microsoft.azuretools.utils.AzureUIRefreshCore;
+import com.microsoft.azuretools.utils.AzureUIRefreshListener;
+import com.microsoft.tooling.msservices.components.DefaultLoader;
+import com.microsoft.tooling.msservices.serviceexplorer.AzureRefreshableNode;
+import com.microsoft.tooling.msservices.serviceexplorer.Node;
+
+import java.io.IOException;
+import java.util.List;
+
+public class FunctionModule extends AzureRefreshableNode implements FunctionModuleView {
+    private static final String FUNCTION_SERVICE_MODULE_ID = FunctionModule.class.getName();
+    private static final String ICON_PATH = "azure-functions-small.png";
+    private static final String BASE_MODULE_NAME = "Function App";
+    private static final String FUNCTION_MODULE = "FunctionModule";
+    private static final String FAILED_TO_DELETE_FUNCTION_APP = "Failed to delete Function App %s";
+    private static final String ERROR_DELETING_FUNCTION_APP = "Azure Services Explorer - Error Deleting Function App";
+    private final FunctionModulePresenter<FunctionModule> functionModulePresenter;
+
+    public FunctionModule(Node parent) {
+        super(FUNCTION_SERVICE_MODULE_ID, BASE_MODULE_NAME, parent, ICON_PATH);
+        functionModulePresenter = new FunctionModulePresenter<>();
+        functionModulePresenter.onAttachView(FunctionModule.this);
+        createListener();
+    }
+
+    @Override
+    protected void refreshItems() throws AzureCmdException {
+        functionModulePresenter.onModuleRefresh();
+    }
+
+    @Override
+    public void removeNode(String sid, String id, Node node) {
+        try {
+            functionModulePresenter.onDeleteFunctionApp(sid, id);
+            removeDirectChildNode(node);
+        } catch (IOException | CloudException e) {
+            DefaultLoader.getUIHelper().showException(String.format(FAILED_TO_DELETE_FUNCTION_APP, node.getName()),
```

Comment:
Showing error stack is not helpful for an UI plugin, you should popup an error notification and then refresh the functions lists.

Reply:
`DefaultLoader.getUIHelper().showException` will prompt a error notification

Reply:
the node is removed but function is not deleted?

Reply:
Exception should be throw in `functionModulePresenter.onDeleteFunctionApp(sid, id);` and catched by exception handler, but in case exception was throw after function was deleted, will still refresh function module when catch exception here